keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/ImageResolver.kt

```diff
@@ -1,75 +1,58 @@
-import com.netflix.spinnaker.keel.api.ec2.ImageProvider
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.LaunchConfigurationSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.ServerGroupSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec.VirtualMachineImage
 import com.netflix.spinnaker.keel.api.ec2.JenkinsImageProvider
 import com.netflix.spinnaker.keel.api.id
-import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
-import com.netflix.spinnaker.keel.clouddriver.DEFAULT_SERVICE_ACCOUNT
 import com.netflix.spinnaker.keel.clouddriver.ImageService
 import com.netflix.spinnaker.keel.clouddriver.model.NamedImage
 import com.netflix.spinnaker.keel.clouddriver.model.appVersion
-import com.netflix.spinnaker.keel.model.SubnetAwareRegionSpec
 import com.netflix.spinnaker.keel.persistence.ArtifactRepository
 import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.plugin.Resolver
 import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import kotlinx.coroutines.runBlocking
 import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
 
+@Component
 class ImageResolver(
   private val dynamicConfigService: DynamicConfigService,
-  private val cloudDriverService: CloudDriverService,
   private val deliveryConfigRepository: DeliveryConfigRepository,
   private val artifactRepository: ArtifactRepository,
   private val imageService: ImageService
-) {
+) : Resolver<ClusterSpec> {
 
-  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  override val apiVersion: ApiVersion = SPINNAKER_API_V1.subApi("ec2")
+  override val supportedKind: String = "cluster"
 
-  // TODO: we may (probably will) want to make the parameter type more generic
-  suspend fun resolveImageId(resource: Resource<ClusterSpec>): ResolvedImages =
-    when (val imageProvider = resource.spec.imageProvider) {
-      is IdImageProvider -> resolveFromImageId(resource, imageProvider)
-      is ArtifactImageProvider -> resolveFromArtifact(resource, imageProvider)
-      is JenkinsImageProvider -> resolveFromJenkinsJob(resource, imageProvider)
-      else -> throw UnsupportedStrategy(
-        imageProvider::class.simpleName.orEmpty(),
-        ImageProvider::class.simpleName.orEmpty()
-      )
-    }
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
 
-  private suspend fun resolveFromImageId(
-    resource: Resource<ClusterSpec>,
-    imageProvider: IdImageProvider
-  ): ResolvedImages {
-    val images = cloudDriverService.namedImages(
-      serviceAccount = DEFAULT_SERVICE_ACCOUNT,
-      imageName = imageProvider.imageId,
-      account = null,
-      region = null
-    )
-    check(images.size == 1) {
-      "Expected a single image for AMI id ${imageProvider.imageId} but found ${images.size}"
+  override fun invoke(resource: Resource<ClusterSpec>): Resource<ClusterSpec> {
+    val imageProvider = resource.spec.imageProvider ?: return resource
```

Comment:
Newbie Kotlin idiom question: if I read this right, here we just return the `resource` as is (i.e. no "resolving") if `resource.spec.imageProvider` is `null`, right? The `else` block if we were writing this in another language would generally have a logging statement in this case, is my guess. Do we just return quietly here because this case doesn't matter much, because it's just the idiom in Kotlin, or something else? 😃

Reply:
If the resource has no `imageProvider` this resolver has no work to do so it just returns the resource unchanged. I'm not sure what logging would add here.